YatimaStdLib/List.lean

```diff
@@ -39,4 +39,83 @@ instance : Foldable List where
   foldr := List.foldr
   foldl := List.foldl
 
+-- the `B` suffix avoids name conflicts with mathlib
+def eraseDupB [BEq α] : List α → List α
+  | [] => []
+  | x::xs => 
+    let exs := eraseDupB xs
+    if exs.contains x then exs else x::exs
+
+def splitAtP [BEq α] (p : α → Bool) (l : List α) : List α × List α :=
+  match l.dropWhile p with 
+  | [] => unreachable!
+  | a::as => ⟨l.takeWhile p ++ [a], as⟩
+
+partial def mergeM [Monad μ] (cmp: α → α → μ Ordering) : List α → List α → μ (List α)
+  | as@(a::as'), bs@(b::bs') => do
+    if (← cmp a b) == Ordering.gt
+    then List.cons b <$> mergeM cmp as bs'
+    else List.cons a <$> mergeM cmp as' bs
+  | [], bs => return bs
+  | as, [] => return as
+
+def mergePairsM [Monad μ] (cmp: α → α → μ Ordering) : List (List α) → μ (List (List α))
+  | a::b::xs => List.cons <$> (mergeM cmp a b) <*> mergePairsM cmp xs
+  | xs => return xs
+
+partial def mergeAllM [Monad μ] (cmp: α → α → μ Ordering) : List (List α) → μ (List α)
+  | [x] => return x
+  | xs => mergePairsM cmp xs >>= mergeAllM cmp
+
+mutual 
+  partial def sequencesM [Monad μ] (cmp : α → α → μ Ordering) : List α → μ (List (List α))
+    | a::b::xs => do
+      if (← cmp a b) == .gt
+      then descendingM cmp b [a] xs 
+      else ascendingM cmp b (fun ys => a :: ys) xs
+    | xs => return [xs]
+
+  partial def descendingM [Monad μ] (cmp : α → α → μ Ordering) (a : α) (as : List α) : List α → μ (List (List α))
+    | b::bs => do
+      if (← cmp a b) == .gt
+      then descendingM cmp b (a::as) bs
+      else List.cons (a::as) <$> sequencesM cmp (b::bs)
+    | [] => List.cons (a::as) <$> sequencesM cmp []
+
+  partial def ascendingM [Monad μ] (cmp : α → α → μ Ordering) (a : α) (as : List α → List α) : List α → μ (List (List α))
+    | b::bs => do
+      if (← cmp a b) != .gt
```

Comment:
Perhaps, `== .lt`, seemingly without the loss of generality? More readable that way.